Authentication

13465012 opened the pull request.

13465012 committed:
Login URL Cosmetics

login page html in place, backend yet to be implemented as still working
on finer points

13465012 committed:
Hash Authentication

* Stores passwords as hashes
* Authenticates login
* Stores login sessions for 1hr in mongodb
* Creates session cookies
* Function to verify logged in

13465012 committed:
Some user authentication implemented

* Editor now requires login
* Domain Specific Authentication
* Index page to access domains after login

13465012 committed:
Introduced index

* Index (home) for selecting domain
* improved login page functionality
* Clicking logo takes user to home page

13465012 committed:
introduced index (continue)

13465012 committed:
Access to pages after new authentication

* Editor access to domain after new login authentication including
- Editor
- View
- History

marc-hanheide commented:
Looks well on track, but I'll wait before I merge until you tell me it's ready.

13465012 committed:
#23 Last Active

* Domains indexed by Last Active
* Landing page list domains by most recently active

13465012 committed:
Domain Permissions Change

* Domains now store users and permissions as nested arrays rather than
as a child object

marc-hanheide commented on app.py:
```diff
@@ -10,23 +10,50 @@
+from web import form
 import signal
 from json import dumps
+import pymongo
```
Do you really need to import the whole module? Can you try to use `from ... import ...` instead for specific objects to import?

marc-hanheide commented on app.py:
```diff
@@ -10,23 +10,50 @@
 import config
 
+# hashing
+import hmac
```
as before: please try to import the minimal set, use `from ... import ...` and then use the code subsequently without the module prefix where possible.

13465012 committed:
#26 Current Features

* All currently implemented features are functional on fresh database
(features added before this commit)

13465012 committed:
#21 Coordinator Level Controls

* Admins, Coordinators and Editors can:
- Edit and Create questions on the domain
- View the Current Question and Stats
- View the history chart for domain

* Admins and Coordinators can:
- Add users to domain (usernames not yet checked to see if they exist)
- Remove users from domain (editor level users)
- View all editors on domain

13465012 committed:
#21 Admin user management

* Admin can now assign roles to users
* Admin gets a list of both editors and coords
* Admin no longer needs to be on the users list for domains having
automatic admin rights instead

13465012 committed:
#20 Longer Session

* Session now 24hr not 1hr

13465012 committed:
#21 Admin Domain Control

* Admin can create  new domains
* Admin can  remove existing domains

13465012 committed:
minimal pymongo imports

* from pymongo import DESCENDING rather than import pymongo

13465012 committed:
#23 Solved Admin Popup Issue

13465012 committed:
#27 Logic Classes and Function

* domain management has its on file
* logging in has its own file
* common imports file
* glob (global variables)
* config moved to modules folder

13465012 committed:
#20 User Management

Admins can now:
* Add new users to the database
* Change the passwords of existing users
* Delete users
* Give users administrative access

Error fixes:
* Only admins can delete domains
* Only users assigned to domains get the link to manage the domain
(before hand they had they link but no ability to manage the domain)

13465012 committed:
#23 Help Button Added

* Help Button added, links to either admin or editor wiki depending on
whether or not you are logged in.
* fixed popup error in home.html where user management button triggers
create new domain popup

13465012 committed:
#27 Code Refactored

* Login Classes in logins.py
* Domain Management in domains.py
* Classes serving URLS in PageServer.py

13465012 committed:
#27 Removed Dead Code

13465012 committed:
Logins Error Fixed

13465012 committed:
Fixed session error

session now overrides rather than return error

13465012 committed:
links use base_url

* Only works if localhost default is same as hosting port (config.py) on
local machine (port 80 on windows for example)
